src/Native/Unix/System.Native/pal_io.c

```diff
@@ -1225,73 +1274,138 @@ int32_t SystemNative_CopyFile(intptr_t sourceFd, intptr_t destinationFd)
         ssize_t sent = sendfile(outFd, inFd, NULL, (size >= SSIZE_MAX ? SSIZE_MAX : (size_t)size));
         if (sent < 0)
         {
-            if (errno != EINVAL && errno != ENOSYS)
-            {
-                return -1;
-            }
-            else
+            if (errno == ENOSYS)
             {
-                break;
+                // If sendfile is unavailable in this kernel, size will
+                // remain untouched, so it's fine for concurrent threads to
+                // write to sendFileUnavailable.
+                sendFileUnavailable = true;
             }
+
+            break;
         }
         else
         {
             assert((size_t)sent <= size);
             size -= (size_t)sent;
         }
     }
-    if (size == 0)
+
+    return size;
+}
+#endif // HAVE_SENDFILE_4
+
+#if defined(PAL_COPY_FILE_RANGE_SYSCALL)
+static loff_t CopyFileRange(int inFd, loff_t *inOff, int outFd, loff_t *outOff, size_t len, unsigned int flags)
+{
+    return syscall(PAL_COPY_FILE_RANGE_SYSCALL, inFd, inOff, outFd, outOff, len, flags);
+}
+
+static uint64_t TryCopyWithCopyFileRange(int inFd, int outFd, const struct stat_ *sourceStat, int64_t startOffset)
+{
+    // As is the case with sendfile(2), copy_file_range(2) will perform an in-kernel
+    // copy of two file descriptors, without copying between userland and the kernel.
+    // It'll try filesystem-specific operations to copy the content, sometimes avoiding
+    // copies altogether (for instance, in NFS shares, copies can happen server-side,
+    // without going back and forth the network).
+    loff_t offIn = startOffset;
+    loff_t offOut = startOffset;
+    size_t toCopy = (size_t)sourceStat->st_size - (size_t)startOffset;
+    static bool copyFileRangeUnavailable = false;
+
+    if (copyFileRangeUnavailable)
     {
-        copied = true;
+        return toCopy;
     }
-    // sendfile couldn't be used; fall back to a manual copy below. This could happen
-    // if we're on an old kernel, for example, where sendfile could only be used
-    // with sockets and not regular files.
-#endif // HAVE_SENDFILE_4
 
-    // Manually read all data from the source and write it to the destination.
-    if (!copied && CopyFile_ReadWrite(inFd, outFd) != 0)
+    while (toCopy) {
+        loff_t copied = CopyFileRange(inFd, &offIn, outFd, &offOut, toCopy, 0);
+
+        // copy_file_range() does not seem to return EINTR according to its man page.
+        if (copied == -1)
+        {
+            if (errno == ENOSYS)
+            {
+                // If copy_file_range is unavailable in this kernel, size
+                // will remain untouched, so it's fine for concurrent
+                // threads to write to sendFileUnavailable.
```

Comment:
I don't understand this comment... what does this check have to do with sendFileUnavailable?  Is that a copy-and-paste error in the comment?